{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "Для того, чтобы загрузить SRA-эксперимент и преобразовать его в .fastq-прочтения, необходимо загрузить SRA Toolkit с сайта NCBI"
      ],
      "metadata": {
        "id": "ANHMEvTYFAyR"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nI36Hamp0zLk",
        "outputId": "91fb62ac-08c9-4246-cf29-5ab9d119b58c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 12:28:42--  https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/3.0.0/sratoolkit.3.0.0-ubuntu64.tar.gz\n",
            "Resolving ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.11, 2607:f220:41f:250::228, ...\n",
            "Connecting to ftp-trace.ncbi.nlm.nih.gov (ftp-trace.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 85777118 (82M) [application/x-gzip]\n",
            "Saving to: ‘sratoolkit.3.0.0-ubuntu64.tar.gz’\n",
            "\n",
            "sratoolkit.3.0.0-ub 100%[===================>]  81.80M  84.2MB/s    in 1.0s    \n",
            "\n",
            "2023-08-06 12:28:43 (84.2 MB/s) - ‘sratoolkit.3.0.0-ubuntu64.tar.gz’ saved [85777118/85777118]\n",
            "\n",
            "sratoolkit.3.0.0-ubuntu64/\n",
            "sratoolkit.3.0.0-ubuntu64/bin/\n",
            "sratoolkit.3.0.0-ubuntu64/bin/abi-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/abi-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/abi-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/abi-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/align-info\n",
            "sratoolkit.3.0.0-ubuntu64/bin/align-info.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/bam-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/bam-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/cache-mgr\n",
            "sratoolkit.3.0.0-ubuntu64/bin/cache-mgr.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/cg-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/cg-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/dump-ref-fasta\n",
            "sratoolkit.3.0.0-ubuntu64/bin/dump-ref-fasta.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fasterq-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fasterq-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fasterq-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/helicos-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/helicos-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/illumina-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/illumina-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/illumina-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/illumina-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/kar\n",
            "sratoolkit.3.0.0-ubuntu64/bin/kar.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/kdbmeta\n",
            "sratoolkit.3.0.0-ubuntu64/bin/kdbmeta.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/latf-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/latf-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/pacbio-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/pacbio-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/prefetch\n",
            "sratoolkit.3.0.0-ubuntu64/bin/prefetch.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/prefetch.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/rcexplain\n",
            "sratoolkit.3.0.0-ubuntu64/bin/rcexplain.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sam-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sam-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sam-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sff-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sff-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sff-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sff-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-pileup\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-pileup.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-pileup.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-search\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-search.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-sort\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-sort-cg\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-sort-cg.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-sort.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-stat\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-stat.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srapath\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srapath.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srapath.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srf-load\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srf-load.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/test-sra\n",
            "sratoolkit.3.0.0-ubuntu64/bin/test-sra.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-config\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-config.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-copy\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-copy.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-decrypt\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-decrypt.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-dump\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-dump.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-encrypt\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-encrypt.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-lock\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-lock.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-unlock\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-unlock.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-validate\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-validate.3\n",
            "sratoolkit.3.0.0-ubuntu64/bin/ncbi/\n",
            "sratoolkit.3.0.0-ubuntu64/bin/ncbi/certs.kfg\n",
            "sratoolkit.3.0.0-ubuntu64/bin/ncbi/default.kfg\n",
            "sratoolkit.3.0.0-ubuntu64/bin/ncbi/vdb-copy.kfg\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srapath-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sratools.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/srf-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/test-sra.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-config.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-copy.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-decrypt.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-dump-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-encrypt.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-lock.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-unlock.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/vdb-validate.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/abi-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/abi-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/align-info.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/bam-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/cache-mgr.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/cg-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/dump-ref-fasta.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fasterq-dump-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-dump-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/fastq-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/helicos-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/illumina-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/illumina-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/kar.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/kdbmeta.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/latf-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/pacbio-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/prefetch-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/rcexplain.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sam-dump-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sff-dump.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sff-load.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-pileup-orig.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-search.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-sort-cg.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-sort.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/bin/sra-stat.3.0.0\n",
            "sratoolkit.3.0.0-ubuntu64/example/\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/base-stats.pl\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/dump-reference.pl\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/gene-lookup.pl\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/mismatch-stats.pl\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/quality-stats.pl\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/simplefastq.pl\n",
            "sratoolkit.3.0.0-ubuntu64/example/perl/splitfastq.pl\n",
            "sratoolkit.3.0.0-ubuntu64/README-blastn\n",
            "sratoolkit.3.0.0-ubuntu64/README-vdb-config\n",
            "sratoolkit.3.0.0-ubuntu64/schema/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/align.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/mate-cache.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/pileup-stats.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/qstat.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/refseq.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/align/seq.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/csra2/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/csra2/csra2.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/csra2/read.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/csra2/reference.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/csra2/stats.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/insdc/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/insdc/insdc.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/insdc/seq.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/insdc/sra.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/clip.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/ncbi.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/pnbrdb.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/seq-graph.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/seq.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/spotname.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/sra.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/stats.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/trace.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/varloc.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/ncbi/wgs-contig.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/454.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/abi.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/generic-fastq.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/helicos.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/illumina.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/ion-torrent.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/nanopore.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/pacbio.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/sra/pevents.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/vdb/\n",
            "sratoolkit.3.0.0-ubuntu64/schema/vdb/built-in.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/schema/vdb/vdb.vschema\n",
            "sratoolkit.3.0.0-ubuntu64/CHANGES\n",
            "sratoolkit.3.0.0-ubuntu64/README.md\n"
          ]
        }
      ],
      "source": [
        "!wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/3.0.0/sratoolkit.3.0.0-ubuntu64.tar.gz\n",
        "!tar -xvzf sratoolkit.3.0.0-ubuntu64.tar.gz"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Необходимо загружаем прочтения при помощи программы fasterq-dump из SRA Toolkit"
      ],
      "metadata": {
        "id": "xfuoWCr4FP17"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sratoolkit.3.0.0-ubuntu64/bin/vdb-config -i"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ktdu2KxoFPmF",
        "outputId": "15ecde11-ce71-4622-9318-0cf407e41c17"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2J\u001b[?25l\u001b[?1000h\u001b[?1002h2023-08-06T12:28:50 vdb-config.3.0.0 fatal: SIGNAL - Segmentation fault \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Теперь запускаем fastq-dump и разархивируем наши прочтения"
      ],
      "metadata": {
        "id": "z-Nw6es2FaMK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sratoolkit.3.0.0-ubuntu64/bin/fasterq-dump SRR7236805 #SRR7236811 SRR7236813 SRR7236815"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nwzoaaz0Fait",
        "outputId": "4e9288e1-e9b2-4fab-b01a-987a750f73e1"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "spots read      : 18,834,191\n",
            "reads read      : 37,668,382\n",
            "reads written   : 18,834,191\n",
            "reads 0-length  : 18,834,191\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Выводим первые и последние несколько строк в прочтениях"
      ],
      "metadata": {
        "id": "13hRlnIBGAIr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print('head')\n",
        "!head -n 4 SRR7236805.fastq\n",
        "print('tail')\n",
        "!tail -n 4 SRR7236805.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KxjUKTtKFz7B",
        "outputId": "f2ccd759-dfaa-43a1-fc43-76843d6ffce6"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "head\n",
            "@SRR7236805.1 1 length=76\n",
            "CTGCANGATAATATCCAGGAGAATTTCCCCAGTCTAGCAAGGCAGGCCAACATTCAAATTCAGGAAATAACAGAGA\n",
            "+SRR7236805.1 1 length=76\n",
            "AA/AA#EEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE/EE\n",
            "tail\n",
            "@SRR7236805.18834191 18834191 length=76\n",
            "TGCTANGTTGACTTAGGGGCTGTGCACAGGAACTAAAAGGCAGGAAAGTACTAAATATTGCTGAGAGCATCCACCC\n",
            "+SRR7236805.18834191 18834191 length=76\n",
            "AAAAA#EEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEAA<\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#print('head')\n",
        "#!head -n 4 SRR7236811.fastq\n",
        "#print('tail')\n",
        "#!tail -n 4 SRR7236811.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zcDU9wvP0iN3",
        "outputId": "aa1d6927-99e6-4898-b6d6-1d6e71790745"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "head\n",
            "@SRR7236811.1 1 length=50\n",
            "NTGCAAATGCAGCGTTAGCAGCTAGTGCAGTAACACAAGTAGCCAAGGTT\n",
            "+SRR7236811.1 1 length=50\n",
            "#4=DDDDFGAFFFHGIJJJIIJJIIHJIIIIHGHIIIIIDHIGIJGFH9D\n",
            "tail\n",
            "@SRR7236811.36396899 36396899 length=49\n",
            "CCAGCCCCTTGGGGCCTCCGTGTTTGNNNNNGGGGAGCTGCTTANNNNN\n",
            "+SRR7236811.36396899 36396899 length=49\n",
            "@@@DDDDDBD>FHI8FHIIIFGFGIH#####00?FAFGEGDGIG#####\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#print('head')\n",
        "#!head -n 4 SRR7236813.fastq\n",
        "#print('tail')\n",
        "#!tail -n 4 SRR7236813.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hCu2ADkz-TiZ",
        "outputId": "8658739b-d5d3-4599-a35e-331b8e33f952"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "head\n",
            "@SRR7236813.1 1 length=50\n",
            "GGGGAGTCAGCCAAGTGCCTCAGTGTGCCCTGTTGAAACTTAGGTTTTTC\n",
            "+SRR7236813.1 1 length=50\n",
            "?<@DBADDH>FHHIIFHIIIIII?FHGIIIIEG<D?DGFGIIIIFCBH<?\n",
            "tail\n",
            "@SRR7236813.46157942 46157942 length=49\n",
            "TTTGTAGTTCAGATATGAGTCATACNNNNNNAAGAGAAAATGTNNNNNN\n",
            "+SRR7236813.46157942 46157942 length=49\n",
            "@@CFDEF;,CFHHIIIIICEHIHHC######21?FHHIIGI@0######\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#print('head')\n",
        "#!head -n 4 SRR7236815.fastq\n",
        "#print('tail')\n",
        "#!tail -n 4 SRR7236815.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Cw_V2BkS-UQn",
        "outputId": "7bf6a109-7a12-4797-beb8-d065d2a6c8f9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "head\n",
            "@SRR7236815.1 HWI-ST984:251:C6YL0ACXX:5:1101:1220:1878 length=50\n",
            "NAGCTCTCGGTCCCCGGCCCCAGCCCAGACCAGGAAGGAGAAAGGAAAAT\n",
            "+SRR7236815.1 HWI-ST984:251:C6YL0ACXX:5:1101:1220:1878 length=50\n",
            "#1=DDFFFHGHHHJJJJJJJJJJJJIGJJJJIIJGJIIFHCGIEHHEHHA\n",
            "tail\n",
            "@SRR7236815.53311340 HWI-ST984:251:C6YL0ACXX:5:2316:21315:100836 length=50\n",
            "CAACCACAGTCACTCCGTGGGCTAAGCGGGGAGCAGGAAGCATGTCTGGT\n",
            "+SRR7236815.53311340 HWI-ST984:251:C6YL0ACXX:5:2316:21315:100836 length=50\n",
            "?+14+A1DDDBBDE?A;+AEADD;EE::?:;5;A@;.5=@(777;>AAA@\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Контроль качества прочтений"
      ],
      "metadata": {
        "id": "uquf_nu0GNAC"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### FastQC"
      ],
      "metadata": {
        "id": "rUAGOeYSGO4G"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Загружаем программу, при помощи которой мы будем проводить оценку качества чтений — FastQC"
      ],
      "metadata": {
        "id": "37z5dh0IGXD-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
        "!unzip fastqc_v0.11.9.zip"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c5dfwy4rGI5P",
        "outputId": "3a12c6ea-c497-467b-9e6d-b57c1905a85b"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 12:31:34--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
            "Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4\n",
            "Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 10249221 (9.8M) [application/zip]\n",
            "Saving to: ‘fastqc_v0.11.9.zip’\n",
            "\n",
            "fastqc_v0.11.9.zip  100%[===================>]   9.77M  2.76MB/s    in 3.5s    \n",
            "\n",
            "2023-08-06 12:31:38 (2.76 MB/s) - ‘fastqc_v0.11.9.zip’ saved [10249221/10249221]\n",
            "\n",
            "Archive:  fastqc_v0.11.9.zip\n",
            "  inflating: FastQC/cisd-jhdf5.jar   \n",
            "   creating: FastQC/Configuration/\n",
            "  inflating: FastQC/Configuration/adapter_list.txt  \n",
            "  inflating: FastQC/Configuration/contaminant_list.txt  \n",
            "  inflating: FastQC/Configuration/limits.txt  \n",
            "  inflating: FastQC/fastqc           \n",
            "  inflating: FastQC/fastqc_icon.ico  \n",
            "   creating: FastQC/Help/\n",
            "   creating: FastQC/Help/1 Introduction/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/entries  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/props/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/text-base/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  \n",
            "   creating: FastQC/Help/2 Basic Operations/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/entries  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/props/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/text-base/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.1 Opening a sequence file.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.2 Evaluating Results.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.3 Saving a Report.html.svn-base  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/2 Basic Operations/2.1 Opening a sequence file.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.2 Evaluating Results.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.3 Saving a Report.html  \n",
            "   creating: FastQC/Help/3 Analysis Modules/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/entries  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/prop-base/\n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/prop-base/kmer_profiles.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_n_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_sequence_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_tile_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/props/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/text-base/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/1 Basic Statistics.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/10 Adapter Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/11 Kmer Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/12 Per Tile Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/2 Per Base Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/3 Per Sequence Quality Scores.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/4 Per Base Sequence Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/5 Per Sequence GC Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/6 Per Base N Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/7 Sequence Length Distribution.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/8 Duplicate Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/9 Overrepresented Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/kmer_profiles.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_n_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_sequence_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_tile_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/1 Basic Statistics.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/10 Adapter Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/11 Kmer Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/12 Per Tile Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/2 Per Base Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/3 Per Sequence Quality Scores.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/4 Per Base Sequence Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/5 Per Sequence GC Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/6 Per Base N Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/7 Sequence Length Distribution.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/8 Duplicate Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/9 Overrepresented Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/duplication_levels.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/kmer_profiles.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_n_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_sequence_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_tile_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/sequence_length_distribution.png  \n",
            "  inflating: FastQC/INSTALL.txt      \n",
            "  inflating: FastQC/jbzip2-0.9.jar   \n",
            "  inflating: FastQC/LICENSE          \n",
            "  inflating: FastQC/LICENSE.txt      \n",
            "  inflating: FastQC/LICENSE_JHDF5.txt  \n",
            "   creating: FastQC/net/\n",
            "   creating: FastQC/net/sourceforge/\n",
            "   creating: FastQC/net/sourceforge/iharder/\n",
            "   creating: FastQC/net/sourceforge/iharder/base64/\n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$1.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$InputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$OutputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64.class  \n",
            "   creating: FastQC/org/\n",
            "   creating: FastQC/org/apache/\n",
            "   creating: FastQC/org/apache/commons/\n",
            "   creating: FastQC/org/apache/commons/math3/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/solvers/\n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AllowedSolution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseAbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BracketedUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BrentSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolverUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/UnivariateFunction.class  \n",
            "   creating: FastQC/org/apache/commons/math3/distribution/\n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractIntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BetaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BinomialDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/CauchyDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ChiSquaredDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/FDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/GammaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/HypergeometricDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/IntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/NormalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PascalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PoissonDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/RealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/SaddlePointExpansion.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/TDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/WeibullDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ZipfDistribution.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/ConvergenceException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/DimensionMismatchException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathArithmeticException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalStateException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathInternalError.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MaxCountExceededException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NoBracketingException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotFiniteNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotStrictlyPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NullArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooLargeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooSmallException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/OutOfRangeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/TooManyEvaluationsException.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ArgUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContext.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContextProvider.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/Localizable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/LocalizedFormats.class  \n",
            "   creating: FastQC/org/apache/commons/math3/random/\n",
            "  inflating: FastQC/org/apache/commons/math3/random/AbstractWell.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/BitsStreamGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomData.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomDataImpl.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/Well19937c.class  \n",
            "   creating: FastQC/org/apache/commons/math3/special/\n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Erf.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma.class  \n",
            "   creating: FastQC/org/apache/commons/math3/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/util/ArithmeticUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ContinuedFraction.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/DoubleArray.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpFracTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpIntTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$lnMant.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathCalc.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathLiteralArrays.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$MaxCountExceededCallback.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/MathUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Precision.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ResizableDoubleArray.class  \n",
            "  inflating: FastQC/README.md        \n",
            "  inflating: FastQC/README.txt       \n",
            "  inflating: FastQC/RELEASE_NOTES.txt  \n",
            "  inflating: FastQC/run_fastqc.bat   \n",
            "  inflating: FastQC/sam-1.103.jar    \n",
            "   creating: FastQC/Templates/\n",
            "  inflating: FastQC/Templates/fastqc2fo.xsl  \n",
            "  inflating: FastQC/Templates/header_template.html  \n",
            "   creating: FastQC/Templates/Icons/\n",
            " extracting: FastQC/Templates/Icons/error.png  \n",
            " extracting: FastQC/Templates/Icons/fastqc_icon.png  \n",
            " extracting: FastQC/Templates/Icons/tick.png  \n",
            " extracting: FastQC/Templates/Icons/warning.png  \n",
            "   creating: FastQC/uk/\n",
            "   creating: FastQC/uk/ac/\n",
            "   creating: FastQC/uk/ac/babraham/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Analysis/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisListener.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisQueue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisRunner.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/OfflineRunner.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Dialogs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel$SmoothJLabel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/WelcomePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCMenuBar.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/FileFilters/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/BAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/CasavaFastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/FastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/GobyFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/MappedBAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/SequenceFileFilter.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Graphs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/BaseGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/LineGraph.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/QualityBoxPlot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/TileGraph.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Help/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot$FileSorter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPage.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay$HelpEditor.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpSearchPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AbstractQCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$Adapter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/DuplicationLevel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModelValue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$Kmer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/NContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$OverrepresentedSeq.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseSequenceContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceGCContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerTileQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/QCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/SequenceLengthDistribution.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Report/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/HTMLReportArchive.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/stylesheet.txt  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Resources/\n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/error.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.png  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.svg  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon_100.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/tick.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/warning.png  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Results/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel$ModuleRenderer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/BAMFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/Contaminant.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminantHit.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminentFinder.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Fast5File.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/FastQFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/PhredEncoding.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Sequence.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFile.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFileGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFormatException.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Statistics/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/NormalDistribution.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/PearsonCorrelation.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Utilities/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/CasavaBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/HotColdColourGradient.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageToBase64.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/MultiMemberGZIPInputStream.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NameFormatException.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NanoporeBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/QualityCount.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/RGB.class  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Делаем скрипт fastqc выполняемым. Выполним анализ качества наших прочтений и положим результаты и все репорты в новосозданную папку qc"
      ],
      "metadata": {
        "id": "ggPIMWb2G_ti"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod +x FastQC/fastqc\n",
        "!mkdir qc\n",
        "!FastQC/fastqc -o qc SRR7236805.fastq #SRR7236811.fastq SRR7236813.fastq SRR7236815.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4eWMjqcI-fxv",
        "outputId": "f9c2be97-ea09-443f-84b6-741512f41d5f"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of SRR7236805.fastq\n",
            "Approx 5% complete for SRR7236805.fastq\n",
            "Approx 10% complete for SRR7236805.fastq\n",
            "Approx 15% complete for SRR7236805.fastq\n",
            "Approx 20% complete for SRR7236805.fastq\n",
            "Approx 25% complete for SRR7236805.fastq\n",
            "Approx 30% complete for SRR7236805.fastq\n",
            "Approx 35% complete for SRR7236805.fastq\n",
            "Approx 40% complete for SRR7236805.fastq\n",
            "Approx 45% complete for SRR7236805.fastq\n",
            "Approx 50% complete for SRR7236805.fastq\n",
            "Approx 55% complete for SRR7236805.fastq\n",
            "Approx 60% complete for SRR7236805.fastq\n",
            "Approx 65% complete for SRR7236805.fastq\n",
            "Approx 70% complete for SRR7236805.fastq\n",
            "Approx 75% complete for SRR7236805.fastq\n",
            "Approx 80% complete for SRR7236805.fastq\n",
            "Approx 85% complete for SRR7236805.fastq\n",
            "Approx 90% complete for SRR7236805.fastq\n",
            "Approx 95% complete for SRR7236805.fastq\n",
            "Analysis complete for SRR7236805.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Для того чтобы посмотреть на .html-репорт, который выдаёт программа FastQC необходимо скачать документ и открыть его"
      ],
      "metadata": {
        "id": "xa2EedwzHwjR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### MultiQC\n",
        "\n",
        "Программа MultiQ необходима для того, чтобы комбинировать несколько различных отчётов в один. Запустим её, дав путь к директории с отчётами FastQC, и посмотрим на .html-отчёт, который будет получен в результате запуска.\n",
        "\n",
        "Получаем html файл и папку."
      ],
      "metadata": {
        "id": "tw9TGUbOH_MW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip3 install multiqc\n",
        "!multiqc qc"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iCeKH-V6IJAx",
        "outputId": "9a40d252-5fd9-46b0-92dd-a97db95c2260"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting multiqc\n",
            "  Downloading multiqc-1.15-py3-none-any.whl (1.3 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m8.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: matplotlib>=2.1.1 in /usr/local/lib/python3.10/dist-packages (from multiqc) (3.7.1)\n",
            "Requirement already satisfied: networkx>=2.5.1 in /usr/local/lib/python3.10/dist-packages (from multiqc) (3.1)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from multiqc) (1.22.4)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from multiqc) (8.1.6)\n",
            "Collecting coloredlogs (from multiqc)\n",
            "  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: future>0.14.0 in /usr/local/lib/python3.10/dist-packages (from multiqc) (0.18.3)\n",
            "Requirement already satisfied: jinja2>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from multiqc) (3.1.2)\n",
            "Collecting lzstring (from multiqc)\n",
            "  Downloading lzstring-1.0.4.tar.gz (4.3 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: markdown in /usr/local/lib/python3.10/dist-packages (from multiqc) (3.4.4)\n",
            "Requirement already satisfied: pyyaml>=4 in /usr/local/lib/python3.10/dist-packages (from multiqc) (6.0.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from multiqc) (2.27.1)\n",
            "Requirement already satisfied: rich>=10 in /usr/local/lib/python3.10/dist-packages (from multiqc) (13.4.2)\n",
            "Collecting rich-click (from multiqc)\n",
            "  Downloading rich_click-1.6.1-py3-none-any.whl (19 kB)\n",
            "Collecting simplejson (from multiqc)\n",
            "  Downloading simplejson-3.19.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (137 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m137.9/137.9 kB\u001b[0m \u001b[31m10.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting spectra>=0.0.10 (from multiqc)\n",
            "  Downloading spectra-0.0.11.tar.gz (18 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2>=3.0.0->multiqc) (2.1.3)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (1.1.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (0.11.0)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (4.41.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (1.4.4)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (23.1)\n",
            "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (9.4.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (3.1.0)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=2.1.1->multiqc) (2.8.2)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10->multiqc) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich>=10->multiqc) (2.14.0)\n",
            "Collecting colormath>=3.0.0 (from spectra>=0.0.10->multiqc)\n",
            "  Downloading colormath-3.0.0.tar.gz (39 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting humanfriendly>=9.1 (from coloredlogs->multiqc)\n",
            "  Downloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m8.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->multiqc) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->multiqc) (2023.7.22)\n",
            "Requirement already satisfied: charset-normalizer~=2.0.0 in /usr/local/lib/python3.10/dist-packages (from requests->multiqc) (2.0.12)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->multiqc) (3.4)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich>=10->multiqc) (0.1.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib>=2.1.1->multiqc) (1.16.0)\n",
            "Building wheels for collected packages: spectra, lzstring, colormath\n",
            "  Building wheel for spectra (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for spectra: filename=spectra-0.0.11-py3-none-any.whl size=17470 sha256=534ea3aee52f9e418455261ae3f3cb39387ae6d6d2105a6974a579d9523d1ad6\n",
            "  Stored in directory: /root/.cache/pip/wheels/bc/27/ec/cddaa9489a49ddcf702e1f4dc1791db28a3b0121e1ff254f97\n",
            "  Building wheel for lzstring (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for lzstring: filename=lzstring-1.0.4-py2.py3-none-any.whl size=4568 sha256=f86376d528ce0364bad74dde79fc1c1ebff4f6fee689ffbe2392ae34ff88f1bf\n",
            "  Stored in directory: /root/.cache/pip/wheels/aa/25/dc/ef10d6c8d762029393d3ef1710c63a7d6714bf1db32d04f3ad\n",
            "  Building wheel for colormath (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for colormath: filename=colormath-3.0.0-py3-none-any.whl size=39407 sha256=bb654ef0ba5e2cbfa023ea79f964ad1e912426cc744df133e2ae7f0079be4bfe\n",
            "  Stored in directory: /root/.cache/pip/wheels/ab/b3/4d/c0738759c25a1df01958068f162cf2a9dc3ab1da8b972cfcfc\n",
            "Successfully built spectra lzstring colormath\n",
            "Installing collected packages: simplejson, lzstring, humanfriendly, colormath, spectra, coloredlogs, rich-click, multiqc\n",
            "Successfully installed coloredlogs-15.0.1 colormath-3.0.0 humanfriendly-10.0 lzstring-1.0.4 multiqc-1.15 rich-click-1.6.1 simplejson-3.19.1 spectra-0.0.11\n",
            "\n",
            "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=334685;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.15\u001b[0m\n",
            "\n",
            "\u001b[34m|           multiqc\u001b[0m | Search path : /content/qc\n",
            "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m2/2\u001b[0m  \n",
            "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 1 reports\n",
            "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc_report.html\n",
            "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc_data\n",
            "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Триммирование\n",
        "Так как у нас в данных присутствуют адаптеры, то необходимо их удалить. Триммирование буду производить только SRR7236805 по адаптеру с помощью trim_galore. Для этого необходимо установить cutadapt.\n",
        "\n",
        "Качество чтений хорошее и картируется >90-95% чтений\n",
        "\n",
        "Далее необходимо еще раз проверить качество оставшихся чтений с помощью fastqc"
      ],
      "metadata": {
        "id": "kmEWNSk0IVpB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install cutadapt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "D1btmP7ROjab",
        "outputId": "25ae8e89-6e79-4e69-cbeb-91def1496b80"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting cutadapt\n",
            "  Downloading cutadapt-4.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (468 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m468.8/468.8 kB\u001b[0m \u001b[31m4.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting dnaio>=0.10.0 (from cutadapt)\n",
            "  Downloading dnaio-0.10.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (81 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m81.8/81.8 kB\u001b[0m \u001b[31m8.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting xopen>=1.6.0 (from cutadapt)\n",
            "  Downloading xopen-1.7.0-py3-none-any.whl (15 kB)\n",
            "Collecting isal>=1.0.0 (from xopen>=1.6.0->cutadapt)\n",
            "  Downloading isal-1.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (250 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m250.7/250.7 kB\u001b[0m \u001b[31m10.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: isal, xopen, dnaio, cutadapt\n",
            "Successfully installed cutadapt-4.4 dnaio-0.10.0 isal-1.2.0 xopen-1.7.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import cutadapt"
      ],
      "metadata": {
        "id": "B4k9TXjpPuom"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir clean\n",
        "!cd /content/clean/\n",
        "!wget https://raw.githubusercontent.com/FelixKrueger/TrimGalore/master/trim_galore\n",
        "!chmod a+x ./trim_galore\n",
        "!./trim_galore"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EOKCO7jHPx94",
        "outputId": "cb819404-6962-49aa-c4a5-0b9d501a2b52"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 12:33:54--  https://raw.githubusercontent.com/FelixKrueger/TrimGalore/master/trim_galore\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 142076 (139K) [text/plain]\n",
            "Saving to: ‘trim_galore’\n",
            "\n",
            "trim_galore         100%[===================>] 138.75K  --.-KB/s    in 0.02s   \n",
            "\n",
            "2023-08-06 12:33:54 (6.21 MB/s) - ‘trim_galore’ saved [142076/142076]\n",
            "\n",
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.4\n",
            "single-core operation.\n",
            "Proceeding with 'gzip' for decompression\n",
            "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "\n",
            "Please provide the filename(s) of one or more FastQ file(s) to launch Trim Galore!\n",
            "\n",
            "USAGE:  'trim_galore [options] <filename(s)>'    or    'trim_galore --help'    for more options\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!/content/trim_galore --help"
      ],
      "metadata": {
        "id": "NtdHm_lZQMyx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Полученные отчеты FastQC и MultiQC просмотрела и данные можно использовать для дальнейшего анализа, единственное, что решила сделать, так это отрезать Overrepresented sequences- TruSeq Adapter, Index 16 (97% over 40bp) в данных SRR7236805.fastq, чтобы попрактиковаться с работой trim_galore. Помимо этого заметила \"Хвостики\" в данных SRR7236811 и в SRR7236813, но их удалять не буду, так как в Hisat2 есть опция по их автоматическому удалению"
      ],
      "metadata": {
        "id": "yemmDiLO_In1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!/content/trim_galore -a \"GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTAT\" --quality 20 --length 30 --output_dir trimmed_data SRR7236805.fastq"
      ],
      "metadata": {
        "id": "oAkCyyeIBFXB",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "74751ee1-55d0-40e9-a005-6657fc882235"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.4\n",
            "single-core operation.\n",
            "Proceeding with 'gzip' for decompression\n",
            "To decrease CPU usage of decompression, please install 'igzip' and run again\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output directory trimmed_data/ doesn't exist, creating it for you...\n",
            "\n",
            "Output will be written into the directory: /content/trimmed_data/\n",
            "Writing report to '/content/trimmed_data/SRR7236805.fastq_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: SRR7236805.fastq\n",
            "Trimming mode: single-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.4\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTAT' (user defined)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length before a sequence gets removed: 30 bp\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.4). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to SRR7236805_trimmed.fq\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTAT' from file SRR7236805.fastq <<< \n",
            "10000000 sequences processed\n",
            "This is cutadapt 4.4 with Python 3.10.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTAT SRR7236805.fastq\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 221.856 s (11.779 µs/read; 5.09 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:              18,834,191\n",
            "Reads with adapters:                 6,259,582 (33.2%)\n",
            "Reads written (passing filters):    18,834,191 (100.0%)\n",
            "\n",
            "Total basepairs processed: 1,431,398,516 bp\n",
            "Quality-trimmed:               6,333,118 bp (0.4%)\n",
            "Total written (filtered):  1,412,003,493 bp (98.6%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: GATCGGAAGAGCACACGTCTGAACTCCAGTCACCCGTCCCGATCTCGTAT; Type: regular 3'; Length: 50; Trimmed: 6259582 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50 bp: 5\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 30.4%\n",
            "  C: 9.8%\n",
            "  G: 29.2%\n",
            "  T: 29.8%\n",
            "  none/other: 0.8%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t4752223\t4708547.8\t0\t4752223\n",
            "2\t1065692\t1177136.9\t0\t1065692\n",
            "3\t269840\t294284.2\t0\t269840\n",
            "4\t60246\t73571.1\t0\t60246\n",
            "5\t8486\t18392.8\t0\t8486\n",
            "6\t4394\t4598.2\t0\t4394\n",
            "7\t2637\t1149.5\t0\t2637\n",
            "8\t2466\t287.4\t0\t2466\n",
            "9\t2886\t71.8\t0\t2340 546\n",
            "10\t2327\t18.0\t1\t1661 666\n",
            "11\t2179\t4.5\t1\t1768 411\n",
            "12\t2395\t1.1\t1\t2045 350\n",
            "13\t1658\t0.3\t1\t1488 170\n",
            "14\t1908\t0.1\t1\t1721 187\n",
            "15\t1091\t0.0\t1\t956 135\n",
            "16\t1435\t0.0\t1\t1256 179\n",
            "17\t1098\t0.0\t1\t966 132\n",
            "18\t972\t0.0\t1\t819 153\n",
            "19\t920\t0.0\t1\t745 171 4\n",
            "20\t942\t0.0\t2\t726 130 86\n",
            "21\t1022\t0.0\t2\t782 139 101\n",
            "22\t905\t0.0\t2\t708 123 74\n",
            "23\t804\t0.0\t2\t637 88 79\n",
            "24\t915\t0.0\t2\t730 117 68\n",
            "25\t679\t0.0\t2\t505 97 77\n",
            "26\t668\t0.0\t2\t513 81 74\n",
            "27\t737\t0.0\t2\t545 99 93\n",
            "28\t541\t0.0\t2\t399 77 65\n",
            "29\t544\t0.0\t2\t392 83 67 2\n",
            "30\t544\t0.0\t3\t356 79 54 55\n",
            "31\t682\t0.0\t3\t464 103 61 54\n",
            "32\t348\t0.0\t3\t198 63 45 42\n",
            "33\t541\t0.0\t3\t373 54 61 53\n",
            "34\t552\t0.0\t3\t358 82 64 48\n",
            "35\t621\t0.0\t3\t391 89 71 70\n",
            "36\t494\t0.0\t3\t320 75 50 49\n",
            "37\t333\t0.0\t3\t211 51 38 32 1\n",
            "38\t407\t0.0\t3\t258 62 40 45 2\n",
            "39\t385\t0.0\t3\t225 67 46 46 1\n",
            "40\t499\t0.0\t4\t247 71 60 57 64\n",
            "41\t795\t0.0\t4\t582 82 55 41 35\n",
            "42\t266\t0.0\t4\t163 38 25 17 23\n",
            "43\t231\t0.0\t4\t136 39 19 17 20\n",
            "44\t37\t0.0\t4\t7 5 9 9 7\n",
            "45\t161\t0.0\t4\t76 27 17 24 17\n",
            "46\t276\t0.0\t4\t145 54 38 17 22\n",
            "47\t285\t0.0\t4\t167 48 30 20 20\n",
            "48\t281\t0.0\t4\t183 42 23 15 16 2\n",
            "49\t186\t0.0\t4\t114 27 16 17 12\n",
            "50\t225\t0.0\t5\t140 37 19 10 10 9\n",
            "51\t259\t0.0\t5\t166 47 17 10 8 11\n",
            "52\t250\t0.0\t5\t153 48 22 14 7 6\n",
            "53\t251\t0.0\t5\t164 50 14 14 4 5\n",
            "54\t367\t0.0\t5\t264 59 24 4 10 6\n",
            "55\t342\t0.0\t5\t257 48 16 11 5 5\n",
            "56\t268\t0.0\t5\t172 53 18 15 5 5\n",
            "57\t243\t0.0\t5\t176 34 14 7 5 7\n",
            "58\t461\t0.0\t5\t348 69 20 10 8 6\n",
            "59\t250\t0.0\t5\t180 37 15 9 4 5\n",
            "60\t252\t0.0\t5\t200 20 5 11 5 11\n",
            "61\t473\t0.0\t5\t365 66 18 10 9 5\n",
            "62\t209\t0.0\t5\t141 25 16 11 11 5\n",
            "63\t280\t0.0\t5\t198 42 12 10 9 9\n",
            "64\t287\t0.0\t5\t206 42 11 14 11 3\n",
            "65\t369\t0.0\t5\t272 52 23 8 5 9\n",
            "66\t382\t0.0\t5\t267 61 22 12 13 7\n",
            "67\t225\t0.0\t5\t151 28 15 14 8 9\n",
            "68\t181\t0.0\t5\t109 34 10 13 7 8\n",
            "69\t176\t0.0\t5\t79 29 17 20 19 12\n",
            "70\t423\t0.0\t5\t273 74 30 21 14 11\n",
            "71\t938\t0.0\t5\t637 169 56 34 23 19\n",
            "72\t2618\t0.0\t5\t2029 386 104 48 28 23\n",
            "73\t6073\t0.0\t5\t5062 752 148 54 34 23\n",
            "74\t27925\t0.0\t5\t25654 1912 223 81 41 14\n",
            "75\t7248\t0.0\t5\t5964 994 147 57 54 32\n",
            "76\t8073\t0.0\t5\t6140 1129 149 154 335 166\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: SRR7236805.fastq\n",
            "=============================================\n",
            "18834191 sequences processed in total\n",
            "Sequences removed because they became shorter than the length cutoff of 30 bp:\t112668 (0.6%)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Снова запустила FastQC для анализа обработанной триммером последовательности"
      ],
      "metadata": {
        "id": "tzAjqBbXA9Wd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir trim_qc\n",
        "!FastQC/fastqc -o trim_qc trimmed_data/SRR7236805_trimmed.fq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Cswz7C5eIt5H",
        "outputId": "21d0aa4f-e90b-41d5-bab5-c1c4c13da1af"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of SRR7236805_trimmed.fq\n",
            "Approx 5% complete for SRR7236805_trimmed.fq\n",
            "Approx 10% complete for SRR7236805_trimmed.fq\n",
            "Approx 15% complete for SRR7236805_trimmed.fq\n",
            "Approx 20% complete for SRR7236805_trimmed.fq\n",
            "Approx 25% complete for SRR7236805_trimmed.fq\n",
            "Approx 30% complete for SRR7236805_trimmed.fq\n",
            "Approx 35% complete for SRR7236805_trimmed.fq\n",
            "Approx 40% complete for SRR7236805_trimmed.fq\n",
            "Approx 45% complete for SRR7236805_trimmed.fq\n",
            "Approx 50% complete for SRR7236805_trimmed.fq\n",
            "Approx 55% complete for SRR7236805_trimmed.fq\n",
            "Approx 60% complete for SRR7236805_trimmed.fq\n",
            "Approx 65% complete for SRR7236805_trimmed.fq\n",
            "Approx 70% complete for SRR7236805_trimmed.fq\n",
            "Approx 75% complete for SRR7236805_trimmed.fq\n",
            "Approx 80% complete for SRR7236805_trimmed.fq\n",
            "Approx 85% complete for SRR7236805_trimmed.fq\n",
            "Approx 90% complete for SRR7236805_trimmed.fq\n",
            "Approx 95% complete for SRR7236805_trimmed.fq\n",
            "Analysis complete for SRR7236805_trimmed.fq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Картирование чтений и анализ выравнивания\n",
        "\n",
        "С помошью программы HISAT2 буду картировать, а затем с помощью пакета SAMTools анализировать данные."
      ],
      "metadata": {
        "id": "4Dz7xRC_q_NN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt install hisat2 samtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sOfSSf50qbhx",
        "outputId": "da315313-8be0-4623-80ab-3179e19f41ea"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  bcftools libhts3 libhtscodecs2 python3-hisat2\n",
            "Suggested packages:\n",
            "  python3-numpy python3-matplotlib texlive-latex-recommended cwltool\n",
            "The following NEW packages will be installed:\n",
            "  bcftools hisat2 libhts3 libhtscodecs2 python3-hisat2 samtools\n",
            "0 upgraded, 6 newly installed, 0 to remove and 15 not upgraded.\n",
            "Need to get 5,505 kB of archives.\n",
            "After this operation, 17.1 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 bcftools amd64 1.13-1 [697 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 hisat2 amd64 2.2.1-3 [3,832 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-hisat2 all 2.2.1-3 [12.7 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 samtools amd64 1.13-4 [520 kB]\n",
            "Fetched 5,505 kB in 1s (3,992 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 6.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package libhtscodecs2:amd64.\n",
            "(Reading database ... 120500 files and directories currently installed.)\n",
            "Preparing to unpack .../0-libhtscodecs2_1.1.1-3_amd64.deb ...\n",
            "Unpacking libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Selecting previously unselected package libhts3:amd64.\n",
            "Preparing to unpack .../1-libhts3_1.13+ds-2build1_amd64.deb ...\n",
            "Unpacking libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Selecting previously unselected package bcftools.\n",
            "Preparing to unpack .../2-bcftools_1.13-1_amd64.deb ...\n",
            "Unpacking bcftools (1.13-1) ...\n",
            "Selecting previously unselected package hisat2.\n",
            "Preparing to unpack .../3-hisat2_2.2.1-3_amd64.deb ...\n",
            "Unpacking hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package python3-hisat2.\n",
            "Preparing to unpack .../4-python3-hisat2_2.2.1-3_all.deb ...\n",
            "Unpacking python3-hisat2 (2.2.1-3) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../5-samtools_1.13-4_amd64.deb ...\n",
            "Unpacking samtools (1.13-4) ...\n",
            "Setting up libhtscodecs2:amd64 (1.1.1-3) ...\n",
            "Setting up libhts3:amd64 (1.13+ds-2build1) ...\n",
            "Setting up bcftools (1.13-1) ...\n",
            "Setting up samtools (1.13-4) ...\n",
            "Setting up hisat2 (2.2.1-3) ...\n",
            "Setting up python3-hisat2 (2.2.1-3) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Буду использовать референс GRCh38"
      ],
      "metadata": {
        "id": "yG4nFSXArCz1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
        "!gzip -d Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
        "!mv Homo_sapiens.GRCh38.dna.primary_assembly.fa genome.fa"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "J7XGvUaVqnK4",
        "outputId": "257d783d-9b36-4fde-c6d2-69355a073aa6"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 12:39:42--  ftp://ftp.ensembl.org/pub/release-84/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\n",
            "           => ‘Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz’\n",
            "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.139\n",
            "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.139|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /pub/release-84/fasta/homo_sapiens/dna ... done.\n",
            "==> SIZE Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz ... 881214344\n",
            "==> PASV ... done.    ==> RETR Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz ... done.\n",
            "Length: 881214344 (840M) (unauthoritative)\n",
            "\n",
            "Homo_sapiens.GRCh38 100%[===================>] 840.39M  29.4MB/s    in 30s     \n",
            "\n",
            "2023-08-06 12:40:13 (28.4 MB/s) - ‘Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz’ saved [881214344]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Создадим файл со справочной информацией о референсной последовательности, в этом нет необходимости, но посмотреть можно"
      ],
      "metadata": {
        "id": "vyHnQPeArT5-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools faidx genome.fa"
      ],
      "metadata": {
        "id": "_nl47movrU8d"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Получим файл с размерами хромосом"
      ],
      "metadata": {
        "id": "I_IUXCShrZ7l"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!awk '{print $1\"\\t\"$2}' genome.fa.fai > genome.chrsize"
      ],
      "metadata": {
        "id": "eOcG_xJkrese"
      },
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Итого, скачала геном человека, разархивировала, проиндексировала, узнала дину хромосомы и имя"
      ],
      "metadata": {
        "id": "9nl3NxPKrhT9"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Индексирую референсную последовательность для картировщика HISAT2"
      ],
      "metadata": {
        "id": "EtODzBY_rjSd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#!hisat2-build --help"
      ],
      "metadata": {
        "id": "FxmKDt27Ln9n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2-build -p 2 genome.fa genome"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "x-p1ha8urDLP",
        "outputId": "c7f27389-9a58-4f90-eafc-1412e3db2a38"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Settings:\n",
            "  Output files: \"genome.*.ht2\"\n",
            "  Line rate: 6 (line is 64 bytes)\n",
            "  Lines per side: 1 (side is 64 bytes)\n",
            "  Offset rate: 4 (one in 16)\n",
            "  FTable chars: 10\n",
            "  Strings: unpacked\n",
            "  Local offset rate: 3 (one in 8)\n",
            "  Local fTable chars: 6\n",
            "  Local sequence length: 57344\n",
            "  Local sequence overlap between two consecutive indexes: 1024\n",
            "  Endianness: little\n",
            "  Actual local endianness: little\n",
            "  Sanity checking: disabled\n",
            "  Assertions: disabled\n",
            "  Random seed: 0\n",
            "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
            "Input files DNA, FASTA:\n",
            "  genome.fa\n",
            "Reading reference sizes\n",
            "  Time reading reference sizes: 00:00:45\n",
            "Calculating joined length\n",
            "Writing header\n",
            "Reserving space for joined string\n",
            "Joining reference sequences\n",
            "  Time to join reference sequences: 00:00:24\n",
            "  Time to read SNPs and splice sites: 00:00:00\n",
            "Using parameters --bmax 276173350 --dcv 1024\n",
            "  Doing ahead-of-time memory usage test\n",
            "  Passed!  Constructing with these parameters: --bmax 276173350 --dcv 1024\n",
            "Constructing suffix-array element generator\n",
            "Building DifferenceCoverSample\n",
            "  Building sPrime\n",
            "  Building sPrimeOrder\n",
            "  V-Sorting samples\n",
            "  V-Sorting samples time: 00:02:51\n",
            "  Allocating rank array\n",
            "  Ranking v-sort output\n",
            "  Ranking v-sort output time: 00:01:03\n",
            "  Invoking Larsson-Sadakane on ranks\n",
            "  Invoking Larsson-Sadakane on ranks time: 00:01:18\n",
            "  Sanity-checking and returning\n",
            "Building samples\n",
            "Reserving space for 22 sample suffixes\n",
            "Generating random suffixes\n",
            "QSorting 22 sample offsets, eliminating duplicates\n",
            "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
            "Multikey QSorting 22 samples\n",
            "  (Using difference cover)\n",
            "  Multikey QSorting samples time: 00:00:00\n",
            "Calculating bucket sizes\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 3, merged 10; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Split 1, merged 3; iterating...\n",
            "Splitting and merging\n",
            "  Splitting and merging time: 00:00:00\n",
            "Avg bucket size: 2.10418e+08 (target: 276173349)\n",
            "Converting suffix-array elements to index image\n",
            "Allocating ftab, absorbFtab\n",
            "Entering GFM loop\n",
            "Getting block 1 of 14\n",
            "  Reserving size (276173350) for bucket 1\n",
            "  Calculating Z arrays for bucket 1\n",
            "Getting block 2 of 14\n",
            "  Entering block accumulator loop for bucket 1:\n",
            "  Reserving size (276173350) for bucket 2\n",
            "  Calculating Z arrays for bucket 2\n",
            "  Entering block accumulator loop for bucket 2:\n",
            "  bucket 1: 10%\n",
            "  bucket 2: 10%\n",
            "  bucket 1: 20%\n",
            "  bucket 2: 20%\n",
            "  bucket 1: 30%\n",
            "  bucket 2: 30%\n",
            "  bucket 1: 40%\n",
            "  bucket 1: 50%\n",
            "  bucket 2: 40%\n",
            "  bucket 1: 60%\n",
            "  bucket 2: 50%\n",
            "  bucket 1: 70%\n",
            "  bucket 2: 60%\n",
            "  bucket 1: 80%\n",
            "  bucket 1: 90%\n",
            "  bucket 2: 70%\n",
            "  bucket 1: 100%\n",
            "  Sorting block of length 267834364 for bucket 1\n",
            "  (Using difference cover)\n",
            "  bucket 2: 80%\n",
            "  bucket 2: 90%\n",
            "  bucket 2: 100%\n",
            "  Sorting block of length 261753040 for bucket 2\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:11:55\n",
            "Returning block of 267834365 for bucket 1\n",
            "Getting block 3 of 14\n",
            "  Reserving size (276173350) for bucket 3\n",
            "  Calculating Z arrays for bucket 3\n",
            "  Entering block accumulator loop for bucket 3:\n",
            "  bucket 3: 10%\n",
            "  Sorting block time: 00:12:14\n",
            "Returning block of 261753041 for bucket 2\n",
            "  bucket 3: 20%\n",
            "  bucket 3: 30%\n",
            "Getting block 4 of 14\n",
            "  Reserving size (276173350) for bucket 4\n",
            "  Calculating Z arrays for bucket 4\n",
            "  Entering block accumulator loop for bucket 4:\n",
            "  bucket 3: 40%\n",
            "  bucket 4: 10%\n",
            "  bucket 3: 50%\n",
            "  bucket 4: 20%\n",
            "  bucket 3: 60%\n",
            "  bucket 4: 30%\n",
            "  bucket 3: 70%\n",
            "  bucket 4: 40%\n",
            "  bucket 3: 80%\n",
            "  bucket 4: 50%\n",
            "  bucket 3: 90%\n",
            "  bucket 3: 100%\n",
            "  Sorting block of length 161973731 for bucket 3\n",
            "  (Using difference cover)\n",
            "  bucket 4: 60%\n",
            "  bucket 4: 70%\n",
            "  bucket 4: 80%\n",
            "  bucket 4: 90%\n",
            "  bucket 4: 100%\n",
            "  Sorting block of length 191484971 for bucket 4\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:09:18\n",
            "Returning block of 161973732 for bucket 3\n",
            "Getting block 5 of 14\n",
            "  Reserving size (276173350) for bucket 5\n",
            "  Calculating Z arrays for bucket 5\n",
            "  Entering block accumulator loop for bucket 5:\n",
            "  bucket 5: 10%\n",
            "  bucket 5: 20%\n",
            "  bucket 5: 30%\n",
            "  bucket 5: 40%\n",
            "  bucket 5: 50%\n",
            "  bucket 5: 60%\n",
            "  bucket 5: 70%\n",
            "  bucket 5: 80%\n",
            "  bucket 5: 90%\n",
            "  bucket 5: 100%\n",
            "  Sorting block of length 233085533 for bucket 5\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:11:10\n",
            "Returning block of 191484972 for bucket 4\n",
            "Getting block 6 of 14\n",
            "  Reserving size (276173350) for bucket 6\n",
            "  Calculating Z arrays for bucket 6\n",
            "  Entering block accumulator loop for bucket 6:\n",
            "  bucket 6: 10%\n",
            "  bucket 6: 20%\n",
            "  bucket 6: 30%\n",
            "  bucket 6: 40%\n",
            "  bucket 6: 50%\n",
            "  bucket 6: 60%\n",
            "  bucket 6: 70%\n",
            "  bucket 6: 80%\n",
            "  bucket 6: 90%\n",
            "  bucket 6: 100%\n",
            "  Sorting block of length 237242639 for bucket 6\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:12:01\n",
            "Returning block of 233085534 for bucket 5\n",
            "Getting block 7 of 14\n",
            "  Reserving size (276173350) for bucket 7\n",
            "  Calculating Z arrays for bucket 7\n",
            "  Entering block accumulator loop for bucket 7:\n",
            "  bucket 7: 10%\n",
            "  bucket 7: 20%\n",
            "  bucket 7: 30%\n",
            "  bucket 7: 40%\n",
            "  bucket 7: 50%\n",
            "  bucket 7: 60%\n",
            "  bucket 7: 70%\n",
            "  bucket 7: 80%\n",
            "  bucket 7: 90%\n",
            "  bucket 7: 100%\n",
            "  Sorting block of length 143059657 for bucket 7\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:12:41\n",
            "Returning block of 237242640 for bucket 6\n",
            "Getting block 8 of 14\n",
            "  Reserving size (276173350) for bucket 8\n",
            "  Calculating Z arrays for bucket 8\n",
            "  Entering block accumulator loop for bucket 8:\n",
            "  bucket 8: 10%\n",
            "  bucket 8: 20%\n",
            "  bucket 8: 30%\n",
            "  bucket 8: 40%\n",
            "  bucket 8: 50%\n",
            "  bucket 8: 60%\n",
            "  bucket 8: 70%\n",
            "  bucket 8: 80%\n",
            "  bucket 8: 90%\n",
            "  bucket 8: 100%\n",
            "  Sorting block of length 251251596 for bucket 8\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:08:16\n",
            "Returning block of 143059658 for bucket 7\n",
            "Getting block 9 of 14\n",
            "  Reserving size (276173350) for bucket 9\n",
            "  Calculating Z arrays for bucket 9\n",
            "  Entering block accumulator loop for bucket 9:\n",
            "  bucket 9: 10%\n",
            "  bucket 9: 20%\n",
            "  bucket 9: 30%\n",
            "  bucket 9: 40%\n",
            "  bucket 9: 50%\n",
            "  bucket 9: 60%\n",
            "  bucket 9: 70%\n",
            "  bucket 9: 80%\n",
            "  bucket 9: 90%\n",
            "  bucket 9: 100%\n",
            "  Sorting block of length 171209757 for bucket 9\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:13:19\n",
            "Returning block of 251251597 for bucket 8\n",
            "Getting block 10 of 14\n",
            "  Reserving size (276173350) for bucket 10\n",
            "  Calculating Z arrays for bucket 10\n",
            "  Entering block accumulator loop for bucket 10:\n",
            "  bucket 10: 10%\n",
            "  bucket 10: 20%\n",
            "  bucket 10: 30%\n",
            "  bucket 10: 40%\n",
            "  bucket 10: 50%\n",
            "  bucket 10: 60%\n",
            "  bucket 10: 70%\n",
            "  bucket 10: 80%\n",
            "  bucket 10: 90%\n",
            "  Sorting block time: 00:08:47\n",
            "Returning block of 171209758 for bucket 9\n",
            "  bucket 10: 100%\n",
            "  Sorting block of length 184703294 for bucket 10\n",
            "  (Using difference cover)\n",
            "Getting block 11 of 14\n",
            "  Reserving size (276173350) for bucket 11\n",
            "  Calculating Z arrays for bucket 11\n",
            "  Entering block accumulator loop for bucket 11:\n",
            "  bucket 11: 10%\n",
            "  bucket 11: 20%\n",
            "  bucket 11: 30%\n",
            "  bucket 11: 40%\n",
            "  bucket 11: 50%\n",
            "  bucket 11: 60%\n",
            "  bucket 11: 70%\n",
            "  bucket 11: 80%\n",
            "  bucket 11: 90%\n",
            "  bucket 11: 100%\n",
            "  Sorting block of length 131915140 for bucket 11\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:06:18\n",
            "Returning block of 131915141 for bucket 11\n",
            "Getting block 12 of 14\n",
            "  Reserving size (276173350) for bucket 12\n",
            "  Calculating Z arrays for bucket 12\n",
            "  Entering block accumulator loop for bucket 12:\n",
            "  bucket 12: 10%\n",
            "  bucket 12: 20%\n",
            "  bucket 12: 30%\n",
            "  Sorting block time: 00:09:50\n",
            "Returning block of 184703295 for bucket 10\n",
            "  bucket 12: 40%\n",
            "  bucket 12: 50%\n",
            "Getting block 13 of 14\n",
            "  Reserving size (276173350) for bucket 13\n",
            "  Calculating Z arrays for bucket 13\n",
            "  Entering block accumulator loop for bucket 13:\n",
            "  bucket 12: 60%\n",
            "  bucket 13: 10%\n",
            "  bucket 12: 70%\n",
            "  bucket 13: 20%\n",
            "  bucket 12: 80%\n",
            "  bucket 13: 30%\n",
            "  bucket 12: 90%\n",
            "  bucket 13: 40%\n",
            "  bucket 12: 100%\n",
            "  Sorting block of length 213313114 for bucket 12\n",
            "  (Using difference cover)\n",
            "  bucket 13: 50%\n",
            "  bucket 13: 60%\n",
            "  bucket 13: 70%\n",
            "  bucket 13: 80%\n",
            "  bucket 13: 90%\n",
            "  bucket 13: 100%\n",
            "  Sorting block of length 274654663 for bucket 13\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:11:06\n",
            "Returning block of 213313115 for bucket 12\n",
            "Getting block 14 of 14\n",
            "  Reserving size (276173350) for bucket 14\n",
            "  Calculating Z arrays for bucket 14\n",
            "  Entering block accumulator loop for bucket 14:\n",
            "  bucket 14: 10%\n",
            "  bucket 14: 20%\n",
            "  bucket 14: 30%\n",
            "  bucket 14: 40%\n",
            "  bucket 14: 50%\n",
            "  bucket 14: 60%\n",
            "  bucket 14: 70%\n",
            "  bucket 14: 80%\n",
            "  bucket 14: 90%\n",
            "  bucket 14: 100%\n",
            "  Sorting block of length 222367555 for bucket 14\n",
            "  (Using difference cover)\n",
            "  Sorting block time: 00:14:44\n",
            "Returning block of 274654664 for bucket 13\n",
            "  Sorting block time: 00:09:58\n",
            "Returning block of 222367556 for bucket 14\n",
            "Exited GFM loop\n",
            "fchr[A]: 0\n",
            "fchr[C]: 869653843\n",
            "fchr[G]: 1470243264\n",
            "fchr[T]: 2073417374\n",
            "fchr[$]: 2945849067\n",
            "Exiting GFM::buildToDisk()\n",
            "Returning from initFromVector\n",
            "Wrote 986172031 bytes to primary GFM file: genome.1.ht2\n",
            "Wrote 736462272 bytes to secondary GFM file: genome.2.ht2\n",
            "Re-opening _in1 and _in2 as input streams\n",
            "Returning from GFM constructor\n",
            "Returning from initFromVector\n",
            "Wrote 1295322177 bytes to primary GFM file: genome.5.ht2\n",
            "Wrote 749943562 bytes to secondary GFM file: genome.6.ht2\n",
            "Re-opening _in5 and _in5 as input streams\n",
            "Returning from HGFM constructor\n",
            "Headers:\n",
            "    len: 2945849067\n",
            "    gbwtLen: 2945849068\n",
            "    nodes: 2945849068\n",
            "    sz: 736462267\n",
            "    gbwtSz: 736462268\n",
            "    lineRate: 6\n",
            "    offRate: 4\n",
            "    offMask: 0xfffffff0\n",
            "    ftabChars: 10\n",
            "    eftabLen: 0\n",
            "    eftabSz: 0\n",
            "    ftabLen: 1048577\n",
            "    ftabSz: 4194308\n",
            "    offsLen: 184115567\n",
            "    offsSz: 736462268\n",
            "    lineSz: 64\n",
            "    sideSz: 64\n",
            "    sideGbwtSz: 48\n",
            "    sideGbwtLen: 192\n",
            "    numSides: 15342964\n",
            "    numLines: 15342964\n",
            "    gbwtTotLen: 981949696\n",
            "    gbwtTotSz: 981949696\n",
            "    reverse: 0\n",
            "    linearFM: Yes\n",
            "Total time for call to driver() for forward index: 02:00:01\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Перемещаю genome.1.ht2 genome.2.ht2 genome.3.ht2 genome.4.ht2 genome.5.ht2 genome.6.ht2 genome.7.ht2 genome.8.ht2 в /content/genome"
      ],
      "metadata": {
        "id": "FEba719Wru0t"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir /content/genome\n",
        "!mv *.ht2 /content/genome"
      ],
      "metadata": {
        "id": "eKVf8r7OiqFv"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#!hisat2 --help"
      ],
      "metadata": {
        "id": "U8arraeWj3SZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Генная разметка"
      ],
      "metadata": {
        "id": "pIiMB-tGr1E9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.annotation.gtf.gz\n",
        "!gunzip gencode.v44.annotation.gtf.gz"
      ],
      "metadata": {
        "id": "mZha7IWzr-uF",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "56ad25ed-287c-49f9-9361-80bce344fcd7"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 16:02:54--  https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/gencode.v44.annotation.gtf.gz\n",
            "Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.138\n",
            "Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.138|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 49721965 (47M) [application/x-gzip]\n",
            "Saving to: ‘gencode.v44.annotation.gtf.gz’\n",
            "\n",
            "gencode.v44.annotat 100%[===================>]  47.42M   643KB/s    in 76s     \n",
            "\n",
            "2023-08-06 16:04:11 (639 KB/s) - ‘gencode.v44.annotation.gtf.gz’ saved [49721965/49721965]\n",
            "\n",
            "gzip: gencode.v44.annotation.gtf already exists; do you wish to overwrite (y or n)? ^C\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!python3 hisat2_extract_splice_sites.py /content/gencode.v44.annotation.gtf > splice_sites.tab"
      ],
      "metadata": {
        "id": "4Y6Z6FyosCVn",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c840eba9-abe4-4a0d-8ca9-b727708f9206"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "python3: can't open file '/content/hisat2_extract_splice_sites.py': [Errno 2] No such file or directory\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!hisat2 --no-softclip -x genome/genome -U /content/trimmed_data/SRR7236805_trimmed.fq -S SRR7236805_trimmed.sam --known-splicesite-infile splice_sites.tab --novel-splicesite-outfile SRR7236805_trimmed.newsplices.tab 2> hisat2.SRR7236805_trimmed.log\n",
        "#!hisat2 -x genome/genome -U /content/SRR7236811.fastq -S SRR7236811.sam --known-splicesite-infile splice_sites.tab --novel-splicesite-outfile SRR7236811.newsplices.tab 2> hisat2.SRR7236811.log\n",
        "#!hisat2 -x genome/genome -U /content/SRR7236813.fastq -S SRR7236813.sam --known-splicesite-infile splice_sites.tab --novel-splicesite-outfile SRR7236813.newsplices.tab 2> hisat2.SRR7236813.log\n",
        "#!hisat2 --no-softclip -x genome/genome -U /content/SRR7236815.fastq -S SRR7236815.sam --known-splicesite-infile splice_sites.tab --novel-splicesite-outfile SRR7236815.newsplices.tab 2> hisat2.SRR7236815.log"
      ],
      "metadata": {
        "id": "6h8vR1VksEAY"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "В файле \"/content/hisat2.SRR7236805_trimmed.log\" указано, сколько чтений картировано на геном: 18721523 reads; of these: 18721523 (100.00%) were unpaired; of these: 514536 (2.75%) aligned 0 times 17035907 (91.00%) aligned exactly 1 time 1171080 (6.26%) aligned >1 times 97.25% overall alignment rate"
      ],
      "metadata": {
        "id": "Sn68gT0YHHaG"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Перевожу sam в bam"
      ],
      "metadata": {
        "id": "Gw0zGJnVHKN9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view SRR7236805_trimmed.sam -o SRR7236805_trimmed.bam\n",
        "#!samtools view SRR7236811.sam -o SRR7236811.bam\n",
        "#!samtools view SRR7236813.sam -o SRR7236813.bam\n",
        "#!samtools view SRR7236815.sam -o SRR7236815.bam"
      ],
      "metadata": {
        "id": "OkqhAEJNsHxu"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/biod/sambamba/releases/download/v0.8.1/sambamba-0.8.1-linux-amd64-static.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "84k9EQP3ycHs",
        "outputId": "d17d41b8-cfd8-4336-8f70-8798012cc7eb"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 15:10:45--  https://github.com/biod/sambamba/releases/download/v0.8.1/sambamba-0.8.1-linux-amd64-static.gz\n",
            "Resolving github.com (github.com)... 140.82.114.3\n",
            "Connecting to github.com (github.com)|140.82.114.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/4167482/c26618cb-f0ee-4190-82b5-b83883afaecf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230806%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230806T151045Z&X-Amz-Expires=300&X-Amz-Signature=00f22246c3e3578078732c73b959bdc20aa09af6890d2bc223884d37382ff8cb&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=4167482&response-content-disposition=attachment%3B%20filename%3Dsambamba-0.8.1-linux-amd64-static.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2023-08-06 15:10:45--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/4167482/c26618cb-f0ee-4190-82b5-b83883afaecf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230806%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230806T151045Z&X-Amz-Expires=300&X-Amz-Signature=00f22246c3e3578078732c73b959bdc20aa09af6890d2bc223884d37382ff8cb&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=4167482&response-content-disposition=attachment%3B%20filename%3Dsambamba-0.8.1-linux-amd64-static.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1704428 (1.6M) [application/octet-stream]\n",
            "Saving to: ‘sambamba-0.8.1-linux-amd64-static.gz’\n",
            "\n",
            "sambamba-0.8.1-linu 100%[===================>]   1.62M  --.-KB/s    in 0.06s   \n",
            "\n",
            "2023-08-06 15:10:45 (28.5 MB/s) - ‘sambamba-0.8.1-linux-amd64-static.gz’ saved [1704428/1704428]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip sambamba-0.8.1-linux-amd64-static.gz"
      ],
      "metadata": {
        "id": "V0qsE2E0mCFV"
      },
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!mv sambamba-0.8.1-linux-amd64-static sambamba"
      ],
      "metadata": {
        "id": "DHx_MAGm0V2f"
      },
      "execution_count": 24,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod a+x sambamba"
      ],
      "metadata": {
        "id": "eRqNcq2B0LOx"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./sambamba"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XUU24KTg0fh8",
        "outputId": "f49d656d-fd4b-4966-88ea-d91fe589720d"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "sambamba 0.8.1\n",
            " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
            "    LDC 1.26.0 / DMD v2.096.1 / LLVM9.0.1 / bootstrap LDC - the LLVM D compiler (0.17.6)\n",
            "\n",
            "\n",
            "Usage: sambamba [command] [args...]\n",
            "\n",
            "  Available commands:\n",
            "\n",
            "    view        view contents and convert from one format\n",
            "                to another (SAM/BAM/JSON/UNPACK)\n",
            "    index       build index (BAI)\n",
            "    merge       merge files (BAM)\n",
            "    sort        sort file (BAM)\n",
            "    slice       slice file (BAM using BED)\n",
            "    markdup     mark or remove duplicates (BAM)\n",
            "    subsample   subsample (BAM)\n",
            "    flagstat    output statistics (BAM)\n",
            "    depth       output statistics (BAM)\n",
            "    validate    simple validator (BAM)\n",
            "\n",
            "  No longer recommended:\n",
            "\n",
            "    mpileup     parallel execution of samtools (BAM)\n",
            "\n",
            "To get help on a particular command, call it without args.\n",
            "\n",
            "Global options\n",
            "\n",
            "    -q          quiet mode (do not show banner)\n",
            "\n",
            "For bug reports and feature requests see\n",
            "\n",
            "       https://github.com/biod/\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools sort SRR7236805_trimmed.bam -o SRR7236805_trimmed.sorted.bam\n",
        "!./sambamba index SRR7236805_trimmed.sorted.bam\n",
        "!samtools depth --reference genome.fa SRR7236805_trimmed.sorted.bam > SRR7236805_trimmed.sorted.depth.tab"
      ],
      "metadata": {
        "id": "xzwJ_EWmsJtG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5a6edfae-e904-4f01-f9f2-059197f1f7a2"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 6 files and 1 in-memory blocks...\n",
            "\n",
            "sambamba 0.8.1\n",
            " by Artem Tarasov and Pjotr Prins (C) 2012-2021\n",
            "    LDC 1.26.0 / DMD v2.096.1 / LLVM9.0.1 / bootstrap LDC - the LLVM D compiler (0.17.6)\n",
            "\n",
            "Usage: samtools depth [options] in.bam [in.bam ...]\n",
            "\n",
            "Options:\n",
            "  -a           Output all positions (including zero depth)\n",
            "  -a -a, -aa   Output absolutely all positions, including unused ref seqs\n",
            "  -r REG       Specify a region in chr or chr:from-to syntax\n",
            "  -b FILE      Use bed FILE for list of regions\n",
            "  -f FILE      Specify list of input BAM/SAM/CRAM filenames\n",
            "  -X           Use custom index files (in -X *.bam *.bam.bai order)\n",
            "  -g INT       Remove specified flags from default flag filter\n",
            "  -G INT       Add specified flags to the default flag filter\n",
            "  -H           Print a file header line\n",
            "  -l INT       Minimum read length [0]\n",
            "  -o FILE      Write output to FILE [stdout]\n",
            "  -q INT       Minimum base quality [0]\n",
            "  -Q INT       Minimum mapping quality [0]\n",
            "  -H           Print a file header\n",
            "  -J           Include reads with deletions in depth computation\n",
            "  -s           Do not count overlapping reads within a template\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -@, --threads INT\n",
            "               Number of additional threads to use [0]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!samtools sort SRR7236811.bam -o SRR7236811.sorted.bam\n",
        "#!sambamba index SRR7236811.sorted.bam\n",
        "#!samtools index SRR7236811.sorted.bam\n",
        "#!samtools depth --reference genome.fa SRR7236811.sorted.bam > SRR7236811.sorted.depth.tab"
      ],
      "metadata": {
        "id": "zj4zMGWfsLxv",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b946f7b8-92f8-45ea-ab71-c0898e767481"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[W::bam_hdr_read] EOF marker is absent. The input is probably truncated\n",
            "[E::bgzf_flush] File write failed (wrong size)\n",
            "[E::bgzf_close] File write failed\n",
            "samtools sort: failed to create temporary file \"SRR7236811.sorted.bam.tmp.0000.bam\": No space left on device\n",
            "/bin/bash: line 1: sambamba: command not found\n",
            "Usage: samtools depth [options] in.bam [in.bam ...]\n",
            "\n",
            "Options:\n",
            "  -a           Output all positions (including zero depth)\n",
            "  -a -a, -aa   Output absolutely all positions, including unused ref seqs\n",
            "  -r REG       Specify a region in chr or chr:from-to syntax\n",
            "  -b FILE      Use bed FILE for list of regions\n",
            "  -f FILE      Specify list of input BAM/SAM/CRAM filenames\n",
            "  -X           Use custom index files (in -X *.bam *.bam.bai order)\n",
            "  -g INT       Remove specified flags from default flag filter\n",
            "  -G INT       Add specified flags to the default flag filter\n",
            "  -H           Print a file header line\n",
            "  -l INT       Minimum read length [0]\n",
            "  -o FILE      Write output to FILE [stdout]\n",
            "  -q INT       Minimum base quality [0]\n",
            "  -Q INT       Minimum mapping quality [0]\n",
            "  -H           Print a file header\n",
            "  -J           Include reads with deletions in depth computation\n",
            "  -s           Do not count overlapping reads within a template\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -@, --threads INT\n",
            "               Number of additional threads to use [0]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!samtools sort SRR7236813.bam -o SRR7236813.sorted.bam\n",
        "#!samtools index SRR7236813.sorted.bam\n",
        "#!samtools depth --reference genome.fa SRR7236813.sorted.bam > SRR7236813.sorted.depth.tab"
      ],
      "metadata": {
        "id": "c_gHfH1St8Aj",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "697d240e-4eee-4b1a-ed35-b29f8befafcc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "samtools sort: failed to read header from \"SRR7236813.bam\"\n",
            "[E::hts_open_format] Failed to open file \"SRR7236813.sorted.bam\" : No such file or directory\n",
            "samtools index: failed to open \"SRR7236813.sorted.bam\": No such file or directory\n",
            "Usage: samtools depth [options] in.bam [in.bam ...]\n",
            "\n",
            "Options:\n",
            "  -a           Output all positions (including zero depth)\n",
            "  -a -a, -aa   Output absolutely all positions, including unused ref seqs\n",
            "  -r REG       Specify a region in chr or chr:from-to syntax\n",
            "  -b FILE      Use bed FILE for list of regions\n",
            "  -f FILE      Specify list of input BAM/SAM/CRAM filenames\n",
            "  -X           Use custom index files (in -X *.bam *.bam.bai order)\n",
            "  -g INT       Remove specified flags from default flag filter\n",
            "  -G INT       Add specified flags to the default flag filter\n",
            "  -H           Print a file header line\n",
            "  -l INT       Minimum read length [0]\n",
            "  -o FILE      Write output to FILE [stdout]\n",
            "  -q INT       Minimum base quality [0]\n",
            "  -Q INT       Minimum mapping quality [0]\n",
            "  -H           Print a file header\n",
            "  -J           Include reads with deletions in depth computation\n",
            "  -s           Do not count overlapping reads within a template\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -@, --threads INT\n",
            "               Number of additional threads to use [0]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!samtools sort SRR7236815.bam -o SRR7236815.sorted.bam\n",
        "#!samtools index SRR7236815.sorted.bam\n",
        "#!samtools depth --reference genome.fa SRR7236815.sorted.bam > SRR7236815.sorted.depth.tab"
      ],
      "metadata": {
        "id": "-F-z8ohqt8ow",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "adf94886-2853-44f9-a2f7-f1d11eece4d3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "samtools sort: failed to read header from \"SRR7236815.bam\"\n",
            "[E::hts_open_format] Failed to open file \"SRR7236815.sorted.bam\" : No such file or directory\n",
            "samtools index: failed to open \"SRR7236815.sorted.bam\": No such file or directory\n",
            "Usage: samtools depth [options] in.bam [in.bam ...]\n",
            "\n",
            "Options:\n",
            "  -a           Output all positions (including zero depth)\n",
            "  -a -a, -aa   Output absolutely all positions, including unused ref seqs\n",
            "  -r REG       Specify a region in chr or chr:from-to syntax\n",
            "  -b FILE      Use bed FILE for list of regions\n",
            "  -f FILE      Specify list of input BAM/SAM/CRAM filenames\n",
            "  -X           Use custom index files (in -X *.bam *.bam.bai order)\n",
            "  -g INT       Remove specified flags from default flag filter\n",
            "  -G INT       Add specified flags to the default flag filter\n",
            "  -H           Print a file header line\n",
            "  -l INT       Minimum read length [0]\n",
            "  -o FILE      Write output to FILE [stdout]\n",
            "  -q INT       Minimum base quality [0]\n",
            "  -Q INT       Minimum mapping quality [0]\n",
            "  -H           Print a file header\n",
            "  -J           Include reads with deletions in depth computation\n",
            "  -s           Do not count overlapping reads within a template\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -@, --threads INT\n",
            "               Number of additional threads to use [0]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "В файле \"/content/hisat2.*.log\" указано, сколько чтений картировано на геном"
      ],
      "metadata": {
        "id": "Lqhn4ED9sPip"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Еще раз посмотрю на bam файл (так как он бинарный используем команду samtools и смотрим на содержимое файла) В результате отображается информация о хромосомах и о их длинах и далее несколько колонон с названием чтения, флагом (https://broadinstitute.github.io/picard/explain-flags.html уникальное-неуникальное картирование/прямое-не прямое картирование и тд)), хромосомах, куда выровнилось чтение, позиция начала выравнивания, скором выравнивания, сигар- (сколько букв выровнилось на референс (М-метч), N-гепы)"
      ],
      "metadata": {
        "id": "bOZEAZ63H8s3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -h /content/SRR7236805_trimmed.sorted.bam | less -S"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "oM9PQfmm22yU",
        "outputId": "befbed99-b7a9-4cf2-e66f-80995826815e"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b7\u001b[?47h\u001b[?1h\u001b=\r@HD     VN:1.0  SO:coordinate\r\n",
            "@SQ     SN:1    LN:248956422\r\n",
            "@SQ     SN:10   LN:133797422\r\n",
            "@SQ     SN:11   LN:135086622\r\n",
            "@SQ     SN:12   LN:133275309\r\n",
            "@SQ     SN:13   LN:114364328\r\n",
            "@SQ     SN:14   LN:107043718\r\n",
            "@SQ     SN:15   LN:101991189\r\n",
            "@SQ     SN:16   LN:90338345\r\n",
            "@SQ     SN:17   LN:83257441\r\n",
            "@SQ     SN:18   LN:80373285\r\n",
            "@SQ     SN:19   LN:58617616\r\n",
            "@SQ     SN:2    LN:242193529\r\n",
            "@SQ     SN:20   LN:64444167\r\n",
            "@SQ     SN:21   LN:46709983\r\n",
            "@SQ     SN:22   LN:50818468\r\n",
            "@SQ     SN:3    LN:198295559\r\n",
            "@SQ     SN:4    LN:190214555\r\n",
            "@SQ     SN:5    LN:181538259\r\n",
            "@SQ     SN:6    LN:170805979\r\n",
            "@SQ     SN:7    LN:159345973\r\n",
            "@SQ     SN:8    LN:145138636\r\n",
            "@SQ     SN:9    LN:138394717\n",
            ":\u001b[K\u0007\u001b[H\u001b[2J\u001b[H\u001b[H\u001b[2J\u001b[H@HD     VN:1.0  SO:coordinate\n",
            "@SQ     SN:1    LN:248956422\n",
            "@SQ     SN:10   LN:133797422\n",
            "@SQ     SN:11   LN:135086622\n",
            "@SQ     SN:12   LN:133275309\n",
            "@SQ     SN:13   LN:114364328\n",
            "@SQ     SN:14   LN:107043718\n",
            "@SQ     SN:15   LN:101991189\n",
            "@SQ     SN:16   LN:90338345\n",
            "@SQ     SN:17   LN:83257441\n",
            "@SQ     SN:18   LN:80373285\n",
            "@SQ     SN:19   LN:58617616\n",
            "@SQ     SN:2    LN:242193529\n",
            "@SQ     SN:20   LN:64444167\n",
            "@SQ     SN:21   LN:46709983\n",
            "@SQ     SN:22   LN:50818468\n",
            "@SQ     SN:3    LN:198295559\n",
            "@SQ     SN:4    LN:190214555\n",
            "@SQ     SN:5    LN:181538259\n",
            "@SQ     SN:6    LN:170805979\n",
            "@SQ     SN:7    LN:159345973\n",
            "@SQ     SN:8    LN:145138636\n",
            "@SQ     SN:9    LN:138394717\n",
            ":\u001b[K\u0007\u001b[H\u001b[2J\u001b[H\u001b[H\u001b[2J\u001b[H@HD     VN:1.0  SO:coordinate\n",
            "@SQ     SN:1    LN:248956422\n",
            "@SQ     SN:10   LN:133797422\n",
            "@SQ     SN:11   LN:135086622\n",
            "@SQ     SN:12   LN:133275309\n",
            "@SQ     SN:13   LN:114364328\n",
            "@SQ     SN:14   LN:107043718\n",
            "@SQ     SN:15   LN:101991189\n",
            "@SQ     SN:16   LN:90338345\n",
            "@SQ     SN:17   LN:83257441\n",
            "@SQ     SN:18   LN:80373285\n",
            "@SQ     SN:19   LN:58617616\n",
            "@SQ     SN:2    LN:242193529\n",
            "@SQ     SN:20   LN:64444167\n",
            "@SQ     SN:21   LN:46709983\n",
            "@SQ     SN:22   LN:50818468\n",
            "@SQ     SN:3    LN:198295559\n",
            "@SQ     SN:4    LN:190214555\n",
            "@SQ     SN:5    LN:181538259\n",
            "@SQ     SN:6    LN:170805979\n",
            "@SQ     SN:7    LN:159345973\n",
            "@SQ     SN:8    LN:145138636\n",
            "@SQ     SN:9    LN:138394717\n",
            ":\u001b[K"
          ]
        },
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-28-bdf2e69ac559>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'samtools view -h /content/SRR7236805_trimmed.sorted.bam | less -S'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_shell.py\u001b[0m in \u001b[0;36msystem\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     97\u001b[0m       \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'also_return_output'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 99\u001b[0;31m     \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_system_commands\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_system_compat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# pylint:disable=protected-access\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    100\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mpip_warn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_system_commands.py\u001b[0m in \u001b[0;36m_system_compat\u001b[0;34m(shell, cmd, also_return_output)\u001b[0m\n\u001b[1;32m    451\u001b[0m   \u001b[0;31m# is expected to call this function, thus adding one level of nesting to the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    452\u001b[0m   \u001b[0;31m# stack.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 453\u001b[0;31m   result = _run_command(\n\u001b[0m\u001b[1;32m    454\u001b[0m       \u001b[0mshell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar_expand\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdepth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mclear_streamed_output\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m   )\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_system_commands.py\u001b[0m in \u001b[0;36m_run_command\u001b[0;34m(cmd, clear_streamed_output)\u001b[0m\n\u001b[1;32m    201\u001b[0m       \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchild_pty\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    202\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 203\u001b[0;31m       \u001b[0;32mreturn\u001b[0m \u001b[0m_monitor_process\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent_pty\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepoll\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mupdate_stdin_widget\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    204\u001b[0m   \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m     \u001b[0mepoll\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_system_commands.py\u001b[0m in \u001b[0;36m_monitor_process\u001b[0;34m(parent_pty, epoll, p, cmd, update_stdin_widget)\u001b[0m\n\u001b[1;32m    231\u001b[0m   \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    232\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 233\u001b[0;31m       \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_poll_process\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent_pty\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepoll\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdecoder\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstate\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    234\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    235\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_system_commands.py\u001b[0m in \u001b[0;36m_poll_process\u001b[0;34m(parent_pty, epoll, p, cmd, decoder, state)\u001b[0m\n\u001b[1;32m    279\u001b[0m   \u001b[0moutput_available\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 281\u001b[0;31m   \u001b[0mevents\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mepoll\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpoll\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    282\u001b[0m   \u001b[0minput_events\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    283\u001b[0m   \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mevent\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mevents\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Чтения, легшие на экзон-интронную границу (поле CIGAR - должно быть N)\n",
        "NH-количество мест картирования, XM-количество ошибокс каким количеством ошибок легло наше чтение), XS-чтения, которые легли с разрывом"
      ],
      "metadata": {
        "id": "W3BksMl_H_Yy"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Подсчёт чтений\n",
        "С помощью набора программ BEDTools был проведён подсчёт чтений в ридах. Команды:"
      ],
      "metadata": {
        "id": "oKu82lJuuTGI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/arq5x/bedtools2/releases/download/v2.28.0/bedtools\n",
        "!chmod a+x ./bedtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-drqka86Ak_r",
        "outputId": "211d2d6e-e3db-4f91-b1cf-7b2b2de64e58"
      },
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-06 16:42:31--  https://github.com/arq5x/bedtools2/releases/download/v2.28.0/bedtools\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/15059334/6cdff700-4d6a-11e9-9ddb-98618909f426?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230806%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230806T164231Z&X-Amz-Expires=300&X-Amz-Signature=3015fa8b55db8903652a18437b6e5ac804e4ba551deb1a21dd6dbbe688f5c41c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools&response-content-type=application%2Foctet-stream [following]\n",
            "--2023-08-06 16:42:31--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/15059334/6cdff700-4d6a-11e9-9ddb-98618909f426?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230806%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230806T164231Z&X-Amz-Expires=300&X-Amz-Signature=3015fa8b55db8903652a18437b6e5ac804e4ba551deb1a21dd6dbbe688f5c41c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 38962368 (37M) [application/octet-stream]\n",
            "Saving to: ‘bedtools’\n",
            "\n",
            "bedtools            100%[===================>]  37.16M  94.6MB/s    in 0.4s    \n",
            "\n",
            "2023-08-06 16:42:32 (94.6 MB/s) - ‘bedtools’ saved [38962368/38962368]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!./bedtools --version"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ff8duSIUJ-d-",
        "outputId": "15e7af3b-22cb-4dc4-c1ae-2f9112f00816"
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "bedtools v2.27.1-65-gc2af1e7-dirty\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!wget -c http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/gtfToGenePred\n",
        "#!wget -c http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/genePredToBed\n",
        "#!chmod 755 gtfToGenePred genePredToBed"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ggLqDxrGdyeL",
        "outputId": "340a95cd-065b-4574-e8ed-7072f7692e92"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2023-08-04 13:24:52--  http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/gtfToGenePred\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 35029160 (33M)\n",
            "Saving to: ‘gtfToGenePred’\n",
            "\n",
            "gtfToGenePred       100%[===================>]  33.41M  21.6MB/s    in 1.5s    \n",
            "\n",
            "2023-08-04 13:24:54 (21.6 MB/s) - ‘gtfToGenePred’ saved [35029160/35029160]\n",
            "\n",
            "--2023-08-04 13:24:54--  http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/genePredToBed\n",
            "Resolving hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)... 128.114.119.163\n",
            "Connecting to hgdownload.cse.ucsc.edu (hgdownload.cse.ucsc.edu)|128.114.119.163|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 35875808 (34M)\n",
            "Saving to: ‘genePredToBed’\n",
            "\n",
            "genePredToBed       100%[===================>]  34.21M  20.2MB/s    in 1.7s    \n",
            "\n",
            "2023-08-04 13:24:56 (20.2 MB/s) - ‘genePredToBed’ saved [35875808/35875808]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Получаю из bam файла выровненные чтения в виде координат (bed)"
      ],
      "metadata": {
        "id": "o-oJCedgIIRn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir bedtool\n",
        "!./bedtools bamtobed -i /content/SRR7236805_trimmed.sorted.bam > bedtool/reads.bed"
      ],
      "metadata": {
        "id": "rPCCejj6IO_A"
      },
      "execution_count": 33,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Получаю bedgraph покрытия генома"
      ],
      "metadata": {
        "id": "yF3WWz5_IvxJ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./bedtools genomecov -ibam /content/SRR7236805_trimmed.sorted.bam -bg | head\n",
        "\n",
        "!./bedtools genomecov -ibam /content/SRR7236805_trimmed.sorted.bam -bga | head"
      ],
      "metadata": {
        "id": "0OYZeKarIwKR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e36ac0a8-31e6-4d4e-cabf-f6bc4fe5631c"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1\t10543\t10619\t1\n",
            "1\t11645\t11720\t1\n",
            "1\t11885\t11886\t1\n",
            "1\t11886\t11960\t2\n",
            "1\t11960\t11961\t1\n",
            "1\t12011\t12085\t1\n",
            "1\t12095\t12170\t1\n",
            "1\t12288\t12363\t1\n",
            "1\t12624\t12700\t1\n",
            "1\t12961\t13036\t1\n",
            "1\t0\t10543\t0\n",
            "1\t10543\t10619\t1\n",
            "1\t10619\t11645\t0\n",
            "1\t11645\t11720\t1\n",
            "1\t11720\t11885\t0\n",
            "1\t11885\t11886\t1\n",
            "1\t11886\t11960\t2\n",
            "1\t11960\t11961\t1\n",
            "1\t11961\t12011\t0\n",
            "1\t12011\t12085\t1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Получить фрагменты генома, покрытые более 50 раз"
      ],
      "metadata": {
        "id": "r2cqkOQEIwfh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./bedtools genomecov -ibam /content/SRR7236805_trimmed.sorted.bam -bg | awk '$4 > 50' | head"
      ],
      "metadata": {
        "id": "1khCumDAIw2C",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9d731cf4-dbae-4d10-e27e-453bffdd8fe3"
      },
      "execution_count": 35,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1\t14479\t14480\t51\n",
            "1\t14480\t14483\t52\n",
            "1\t14483\t14484\t56\n",
            "1\t14484\t14485\t57\n",
            "1\t14485\t14486\t53\n",
            "1\t14486\t14489\t54\n",
            "1\t14489\t14492\t60\n",
            "1\t14492\t14493\t61\n",
            "1\t14493\t14494\t62\n",
            "1\t14494\t14497\t64\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Аннотация"
      ],
      "metadata": {
        "id": "dhmbqvDMd_X_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt install python3-htseq"
      ],
      "metadata": {
        "id": "BO8X0E5Au-72",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d9a71127-5320-4ee0-d9b0-9476066adf37"
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  python3-numpy python3-pysam\n",
            "Suggested packages:\n",
            "  python-numpy-doc python3-pytest\n",
            "The following NEW packages will be installed:\n",
            "  python3-htseq python3-numpy python3-pysam\n",
            "0 upgraded, 3 newly installed, 0 to remove and 15 not upgraded.\n",
            "Need to get 5,499 kB of archives.\n",
            "After this operation, 25.7 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 python3-numpy amd64 1:1.21.5-1ubuntu22.04.1 [3,467 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-pysam amd64 0.17.0+ds-2build1 [1,749 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 python3-htseq amd64 1.99.2-1build1 [283 kB]\n",
            "Fetched 5,499 kB in 1s (8,514 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 3.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package python3-numpy.\n",
            "(Reading database ... 120700 files and directories currently installed.)\n",
            "Preparing to unpack .../python3-numpy_1%3a1.21.5-1ubuntu22.04.1_amd64.deb ...\n",
            "Unpacking python3-numpy (1:1.21.5-1ubuntu22.04.1) ...\n",
            "Selecting previously unselected package python3-pysam.\n",
            "Preparing to unpack .../python3-pysam_0.17.0+ds-2build1_amd64.deb ...\n",
            "Unpacking python3-pysam (0.17.0+ds-2build1) ...\n",
            "Selecting previously unselected package python3-htseq.\n",
            "Preparing to unpack .../python3-htseq_1.99.2-1build1_amd64.deb ...\n",
            "Unpacking python3-htseq (1.99.2-1build1) ...\n",
            "Setting up python3-pysam (0.17.0+ds-2build1) ...\n",
            "Setting up python3-numpy (1:1.21.5-1ubuntu22.04.1) ...\n",
            "Setting up python3-htseq (1.99.2-1build1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "С помощью htseq-count для каждого гена посчитаем число попавших на него чтений"
      ],
      "metadata": {
        "id": "ncI-orTyeXmW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir htseq\n",
        "!htseq-count -f bam -t gene -s no -m union /content/SRR7236805_trimmed.sorted.bam /content/gencode.v44.annotation.gtf > htseq/SRR7236805.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Qt01PrjleDWo",
        "outputId": "439b5b6a-3460-4a40-b3ce-b74e409f24fb"
      },
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "100000 GFF lines processed.\n",
            "200000 GFF lines processed.\n",
            "300000 GFF lines processed.\n",
            "400000 GFF lines processed.\n",
            "500000 GFF lines processed.\n",
            "600000 GFF lines processed.\n",
            "700000 GFF lines processed.\n",
            "800000 GFF lines processed.\n",
            "900000 GFF lines processed.\n",
            "1000000 GFF lines processed.\n",
            "1100000 GFF lines processed.\n",
            "1200000 GFF lines processed.\n",
            "1300000 GFF lines processed.\n",
            "1400000 GFF lines processed.\n",
            "1500000 GFF lines processed.\n",
            "1600000 GFF lines processed.\n",
            "1700000 GFF lines processed.\n",
            "1800000 GFF lines processed.\n",
            "1900000 GFF lines processed.\n",
            "2000000 GFF lines processed.\n",
            "2100000 GFF lines processed.\n",
            "2200000 GFF lines processed.\n",
            "2300000 GFF lines processed.\n",
            "2400000 GFF lines processed.\n",
            "2500000 GFF lines processed.\n",
            "2600000 GFF lines processed.\n",
            "2700000 GFF lines processed.\n",
            "2800000 GFF lines processed.\n",
            "2900000 GFF lines processed.\n",
            "3000000 GFF lines processed.\n",
            "3100000 GFF lines processed.\n",
            "3200000 GFF lines processed.\n",
            "3300000 GFF lines processed.\n",
            "3400000 GFF lines processed.\n",
            "3424189 GFF lines processed.\n",
            "100000 alignment records processed.\n",
            "200000 alignment records processed.\n",
            "300000 alignment records processed.\n",
            "400000 alignment records processed.\n",
            "500000 alignment records processed.\n",
            "600000 alignment records processed.\n",
            "700000 alignment records processed.\n",
            "800000 alignment records processed.\n",
            "900000 alignment records processed.\n",
            "1000000 alignment records processed.\n",
            "1100000 alignment records processed.\n",
            "1200000 alignment records processed.\n",
            "1300000 alignment records processed.\n",
            "1400000 alignment records processed.\n",
            "1500000 alignment records processed.\n",
            "1600000 alignment records processed.\n",
            "1700000 alignment records processed.\n",
            "1800000 alignment records processed.\n",
            "1900000 alignment records processed.\n",
            "2000000 alignment records processed.\n",
            "2100000 alignment records processed.\n",
            "2200000 alignment records processed.\n",
            "2300000 alignment records processed.\n",
            "2400000 alignment records processed.\n",
            "2500000 alignment records processed.\n",
            "2600000 alignment records processed.\n",
            "2700000 alignment records processed.\n",
            "2800000 alignment records processed.\n",
            "2900000 alignment records processed.\n",
            "3000000 alignment records processed.\n",
            "3100000 alignment records processed.\n",
            "3200000 alignment records processed.\n",
            "3300000 alignment records processed.\n",
            "3400000 alignment records processed.\n",
            "3500000 alignment records processed.\n",
            "3600000 alignment records processed.\n",
            "3700000 alignment records processed.\n",
            "3800000 alignment records processed.\n",
            "3900000 alignment records processed.\n",
            "4000000 alignment records processed.\n",
            "4100000 alignment records processed.\n",
            "4200000 alignment records processed.\n",
            "4300000 alignment records processed.\n",
            "4400000 alignment records processed.\n",
            "4500000 alignment records processed.\n",
            "4600000 alignment records processed.\n",
            "4700000 alignment records processed.\n",
            "4800000 alignment records processed.\n",
            "4900000 alignment records processed.\n",
            "5000000 alignment records processed.\n",
            "5100000 alignment records processed.\n",
            "5200000 alignment records processed.\n",
            "5300000 alignment records processed.\n",
            "5400000 alignment records processed.\n",
            "5500000 alignment records processed.\n",
            "5600000 alignment records processed.\n",
            "5700000 alignment records processed.\n",
            "5800000 alignment records processed.\n",
            "5900000 alignment records processed.\n",
            "6000000 alignment records processed.\n",
            "6100000 alignment records processed.\n",
            "6200000 alignment records processed.\n",
            "6300000 alignment records processed.\n",
            "6400000 alignment records processed.\n",
            "6500000 alignment records processed.\n",
            "6600000 alignment records processed.\n",
            "6700000 alignment records processed.\n",
            "6800000 alignment records processed.\n",
            "6900000 alignment records processed.\n",
            "7000000 alignment records processed.\n",
            "7100000 alignment records processed.\n",
            "7200000 alignment records processed.\n",
            "7300000 alignment records processed.\n",
            "7400000 alignment records processed.\n",
            "7500000 alignment records processed.\n",
            "7600000 alignment records processed.\n",
            "7700000 alignment records processed.\n",
            "7800000 alignment records processed.\n",
            "7900000 alignment records processed.\n",
            "8000000 alignment records processed.\n",
            "8100000 alignment records processed.\n",
            "8200000 alignment records processed.\n",
            "8300000 alignment records processed.\n",
            "8400000 alignment records processed.\n",
            "8500000 alignment records processed.\n",
            "8600000 alignment records processed.\n",
            "8700000 alignment records processed.\n",
            "8800000 alignment records processed.\n",
            "8900000 alignment records processed.\n",
            "9000000 alignment records processed.\n",
            "9100000 alignment records processed.\n",
            "9200000 alignment records processed.\n",
            "9300000 alignment records processed.\n",
            "9400000 alignment records processed.\n",
            "9500000 alignment records processed.\n",
            "9600000 alignment records processed.\n",
            "9700000 alignment records processed.\n",
            "9800000 alignment records processed.\n",
            "9900000 alignment records processed.\n",
            "10000000 alignment records processed.\n",
            "10100000 alignment records processed.\n",
            "10200000 alignment records processed.\n",
            "10300000 alignment records processed.\n",
            "10400000 alignment records processed.\n",
            "10500000 alignment records processed.\n",
            "10600000 alignment records processed.\n",
            "10700000 alignment records processed.\n",
            "10800000 alignment records processed.\n",
            "10900000 alignment records processed.\n",
            "11000000 alignment records processed.\n",
            "11100000 alignment records processed.\n",
            "11200000 alignment records processed.\n",
            "11300000 alignment records processed.\n",
            "11400000 alignment records processed.\n",
            "11500000 alignment records processed.\n",
            "11600000 alignment records processed.\n",
            "11700000 alignment records processed.\n",
            "11800000 alignment records processed.\n",
            "11900000 alignment records processed.\n",
            "12000000 alignment records processed.\n",
            "12100000 alignment records processed.\n",
            "12200000 alignment records processed.\n",
            "12300000 alignment records processed.\n",
            "12400000 alignment records processed.\n",
            "12500000 alignment records processed.\n",
            "12600000 alignment records processed.\n",
            "12700000 alignment records processed.\n",
            "12800000 alignment records processed.\n",
            "12900000 alignment records processed.\n",
            "13000000 alignment records processed.\n",
            "13100000 alignment records processed.\n",
            "13200000 alignment records processed.\n",
            "13300000 alignment records processed.\n",
            "13400000 alignment records processed.\n",
            "13500000 alignment records processed.\n",
            "13600000 alignment records processed.\n",
            "13700000 alignment records processed.\n",
            "13800000 alignment records processed.\n",
            "13900000 alignment records processed.\n",
            "14000000 alignment records processed.\n",
            "14100000 alignment records processed.\n",
            "14200000 alignment records processed.\n",
            "14300000 alignment records processed.\n",
            "14400000 alignment records processed.\n",
            "14500000 alignment records processed.\n",
            "14600000 alignment records processed.\n",
            "14700000 alignment records processed.\n",
            "14800000 alignment records processed.\n",
            "14900000 alignment records processed.\n",
            "15000000 alignment records processed.\n",
            "15100000 alignment records processed.\n",
            "15200000 alignment records processed.\n",
            "15300000 alignment records processed.\n",
            "15400000 alignment records processed.\n",
            "15500000 alignment records processed.\n",
            "15600000 alignment records processed.\n",
            "15700000 alignment records processed.\n",
            "15800000 alignment records processed.\n",
            "15900000 alignment records processed.\n",
            "16000000 alignment records processed.\n",
            "16100000 alignment records processed.\n",
            "16200000 alignment records processed.\n",
            "16300000 alignment records processed.\n",
            "16400000 alignment records processed.\n",
            "16500000 alignment records processed.\n",
            "16600000 alignment records processed.\n",
            "16700000 alignment records processed.\n",
            "16800000 alignment records processed.\n",
            "16900000 alignment records processed.\n",
            "17000000 alignment records processed.\n",
            "17100000 alignment records processed.\n",
            "17200000 alignment records processed.\n",
            "17300000 alignment records processed.\n",
            "17400000 alignment records processed.\n",
            "17500000 alignment records processed.\n",
            "17600000 alignment records processed.\n",
            "17700000 alignment records processed.\n",
            "17800000 alignment records processed.\n",
            "17900000 alignment records processed.\n",
            "18000000 alignment records processed.\n",
            "18100000 alignment records processed.\n",
            "18200000 alignment records processed.\n",
            "18300000 alignment records processed.\n",
            "18400000 alignment records processed.\n",
            "18500000 alignment records processed.\n",
            "18600000 alignment records processed.\n",
            "18700000 alignment records processed.\n",
            "18800000 alignment records processed.\n",
            "18900000 alignment records processed.\n",
            "19000000 alignment records processed.\n",
            "19100000 alignment records processed.\n",
            "19200000 alignment records processed.\n",
            "19300000 alignment records processed.\n",
            "19400000 alignment records processed.\n",
            "19500000 alignment records processed.\n",
            "19600000 alignment records processed.\n",
            "19700000 alignment records processed.\n",
            "19800000 alignment records processed.\n",
            "19900000 alignment records processed.\n",
            "20000000 alignment records processed.\n",
            "20100000 alignment records processed.\n",
            "20200000 alignment records processed.\n",
            "20300000 alignment records processed.\n",
            "20400000 alignment records processed.\n",
            "20500000 alignment records processed.\n",
            "20600000 alignment records processed.\n",
            "20700000 alignment records processed.\n",
            "20800000 alignment records processed.\n",
            "20842076 alignment records processed.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Входной файл в формате bam, смотрим гены, библиотека не цепьспецифичная, параметр union"
      ],
      "metadata": {
        "id": "ounkOBCsffhZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!head htseq/SRR7236805.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bMF23IqchdbK",
        "outputId": "2a702f8a-f3b5-4dae-f2e2-13a4095297d0"
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ENSG00000000003.16\t0\n",
            "ENSG00000000005.6\t0\n",
            "ENSG00000000419.14\t0\n",
            "ENSG00000000457.14\t0\n",
            "ENSG00000000460.17\t0\n",
            "ENSG00000000938.13\t0\n",
            "ENSG00000000971.17\t0\n",
            "ENSG00000001036.14\t0\n",
            "ENSG00000001084.13\t0\n",
            "ENSG00000001167.15\t0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tail htseq/SRR7236805.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iJZ6icQRhha6",
        "outputId": "71a1b6f0-1923-43e8-8b08-4f5410d870b1"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ENSG00000292369.1\t0\n",
            "ENSG00000292370.1\t0\n",
            "ENSG00000292371.1\t0\n",
            "ENSG00000292372.1\t0\n",
            "ENSG00000292373.1\t0\n",
            "__no_feature\t17035907\n",
            "__ambiguous\t0\n",
            "__too_low_aQual\t0\n",
            "__not_aligned\t514536\n",
            "__alignment_not_unique\t1171080\n"
          ]
        }
      ]
    }
  ]
}